File: hittable_list.c
#include <stdbool.h>
#include "hittable_list.h"
#include "sphere.h"
#include "hittable.h"

bool pool_hits(HittableList list, Ray r, float t_min, float t_max, HitRecord *rec)
{
    HitRecord temp_rec;
    HitRecord *temp_rec_p = &temp_rec;

    bool hit_anything = false;
    double closest_so_far = t_max;
    for (int i = 0; i < list.size; i++)
    {
        if (hit(list.objects[i], r, t_min, t_max, temp_rec_p))
        {
            hit_anything = true;
            closest_so_far = temp_rec.t;
            *rec = temp_rec;
        };
    }
    return hit_anything;
};

bool object_hit(Object object, Ray r, float t_min, float t_max, HitRecord *rec)
{
    switch (object.type)
    {
    case sphere:
        Sphere sph = object.data.s;
        return sphere_hit(sph, r, t_min, t_max, *rec);
    }
}

// HittableList hittable_list_append(HittableList list, Object *object)
// {
//     list.size += 1;
//     Object object_list[list.size];
//     for (int i = 0; i < list.size - 1; i++)
//     {
//         object_list[i] = list.objects[i];
//     }
//     object_list[list.size - 1] = object;

//     HittableList new_list = {
//         .size = list.size,
//         .objects = object_list};
// }

File: python/hittable_list.py
from hittable import Hittable, HitRecord
import numpy as np
from ray import Ray
from typing import List


class HittableList(Hittable):
    def __init__(self, objects: List[Hittable]):
        self.objects = objects

    def hit(self, r: Ray, t_min: float, t_max: float, rec: HitRecord):
        temp_rec = HitRecord(0.0, np.ndarray([]), np.ndarray([]))
        hit_anything = False
        closest_so_far = t_max
        for object in self.objects:
            if object.hit(r, t_min, closest_so_far, temp_rec):
                hit_anything = True
                closest_so_far = temp_rec.t
                rec.t = temp_rec.t
                rec.p = temp_rec.p
                rec.normal = temp_rec.normal
        return hit_anything
